nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/cosmos/document/AzureCosmosDBConnectionControllerService.java

```diff
@@ -0,0 +1,188 @@
+    static final String CONSISTENCY_EVENTUAL = "EVENTUAL";
+    private String uri;
+    private String accessKey;
+    protected CosmosClient cosmosClient;
+
+    public static final PropertyDescriptor URI = new PropertyDescriptor.Builder()
+        .name("azure-cosmos-db-uri")
+        .displayName("Azure Cosmos DB URI")
+        .description("CosmosURI, typically of the form: https://{databaseaccount}.documents.azure.com:443/. Note this URI is for Azure Cosmos DB with SQL API")
+        .required(true)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    public static final PropertyDescriptor DB_ACCESS_KEY = new PropertyDescriptor.Builder()
+        .name("azure-cosmos-db-key")
+        .displayName("Azure Cosmos DB Access Key")
+        .description("Azure Cosmos DB Access Key from Azure Portal (Settings->Keys)")
+        .required(true)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .sensitive(true)
+        .build();
+
+    public static final PropertyDescriptor CONSISTENCY = new PropertyDescriptor.Builder()
+        .name("azure-cosmos-consistency-level")
+        .displayName("Azure Cosmos DB Consistency Level")
+        .description("Azure Cosmos DB Consistency Level to use")
+        .required(false)
+        .allowableValues(CONSISTENCY_STRONG, CONSISTENCY_BOUNDED_STALENESS, CONSISTENCY_SESSION,
+                CONSISTENCY_CONSISTENT_PREFIX, CONSISTENCY_EVENTUAL)
+        .defaultValue(CONSISTENCY_SESSION)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .build();
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        this.uri = context.getProperty(URI).getValue();
+        this.accessKey = context.getProperty(DB_ACCESS_KEY).getValue();
+        final ConsistencyLevel clevel;
+        final String selectedConsistency = context.getProperty(CONSISTENCY).getValue();
+
+        switch(selectedConsistency) {
+            case CONSISTENCY_STRONG:
+                clevel =  ConsistencyLevel.STRONG;
+                break;
+            case CONSISTENCY_CONSISTENT_PREFIX:
+                clevel = ConsistencyLevel.CONSISTENT_PREFIX;
+                break;
+            case CONSISTENCY_SESSION:
+                clevel = ConsistencyLevel.SESSION;
+                break;
+            case CONSISTENCY_BOUNDED_STALENESS:
+                clevel = ConsistencyLevel.BOUNDED_STALENESS;
+                break;
+            case CONSISTENCY_EVENTUAL:
+                clevel = ConsistencyLevel.EVENTUAL;
+                break;
+            default:
+                clevel = ConsistencyLevel.SESSION;
+        }
+
+        if (cosmosClient != null) {
+            closeClient();
+        }
+        createCosmosClient(uri, accessKey, clevel);
+    }
+
+
+    @OnStopped
+    public final void closeClient() {
+        if (this.cosmosClient != null) {
+            cosmosClient.close();
```

Comment:
Does this ever throw any exceptions? If so then I think it should get wrapped in a try-catch ignore so `cosmosClient = null` definitely happens.

Reply:
done.